File: sdk-js/src/language-model.ts
import type {
  LanguageModelCapability,
  LanguageModelInput,
  LanguageModelPricing,
  ModelResponse,
  PartialModelResponse,
} from "./types.ts";

export interface LanguageModelMetadata {
  /**
   * The pricing per single token for the model. Used to calculate the cost of a response.
   */
  pricing?: LanguageModelPricing;
  /**
   * The capabilities of the model.
   */
  capabilities?: LanguageModelCapability[];
}

export interface LanguageModel {
  /**
   * The provider of the model, e.g. "openai", "anthropic", "google"
   */
  provider: string;
  /**
   * The ID of the model, e.g. "gpt-3.5-turbo", "haiku"
   */
  modelId: string;
  /**
   * The metadata of the model.
   */
  metadata?: LanguageModelMetadata;
  /**
   * Generates a response to the given input.
   */
  generate(input: LanguageModelInput): Promise<ModelResponse>;
  /**
   * Generates a response to the given input, returning a stream of partial responses.
   */
  stream(input: LanguageModelInput): AsyncGenerator<PartialModelResponse>;
}

File: sdk-go/languagemodel.go
package llmsdk

import (
	"context"

	"github.com/hoangvvo/llm-sdk/sdk-go/utils/stream"
)

type ProviderName string

type LanguageModelMetadata struct {
	Pricing      *LanguageModelPricing     `json:"pricing"`
	Capabilities []LanguageModelCapability `json:"capabilities,omitempty"`
}

type LanguageModel interface {
	Provider() ProviderName
	ModelID() string
	Metadata() *LanguageModelMetadata
	Generate(ctx context.Context, input *LanguageModelInput) (*ModelResponse, error)
	Stream(ctx context.Context, input *LanguageModelInput) (*LanguageModelStream, error)
}

type LanguageModelStream = stream.Stream[*PartialModelResponse]
